src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs

```diff
@@ -40,5 +40,54 @@ public static unsafe bool TryGetRawMetadata(this Assembly assembly, out byte* bl
 
             return InternalTryGetRawMetadata(new QCallAssembly(ref rtAsm), ref blob, ref length);
         }
+
+        [DllImport(RuntimeHelpers.QCall)]
+        private static extern unsafe int ApplyHotReloadUpdate(QCallAssembly assembly, byte* metadataDelta, int metadataDeltaLength, byte* ilDelta, int ilDeltaLength, byte* pdbDelta, int pdbDeltaLength);
+
+        /// <summary>
+        /// Hot reload update API
+        /// Applies an update to the given assembly using the metadata, IL and PDB deltas. Currently executing
+        /// methods will continue to use the existing IL. New executions of modified methods will use the new
+        /// IL. The supported changes are runtime specific - different .NET runtimes may have different
+        /// limitations. The runtime makes no guarantees if the delta includes unsupported changes.
+        /// </summary>
+        /// <param name="assembly">The assembly to update</param>
+        /// <param name="metadataDelta">The metadata changes</param>
+        /// <param name="ilDelta">The IL changes</param>
+        /// <param name="pdbDelta">The PDB changes. Current not supported on .NET Core</param>
+        /// <exception cref="ArgumentNullException">if assembly parameter is null</exception>
+        /// <exception cref="NotSupportedException">update failed</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta = default)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            RuntimeAssembly? runtimeAssembly = assembly as RuntimeAssembly;
+            if (runtimeAssembly == null)
+            {
+                throw new ArgumentException("Not a RuntimeAssembly", nameof(assembly));
+            }
+
+            unsafe
+            {
+                RuntimeAssembly rtAsm = runtimeAssembly;
+                fixed (byte* metadataDeltaPtr = metadataDelta, ilDeltaPtr = ilDelta, pdbDeltaPtr = pdbDelta)
+                {
+                    if (ApplyHotReloadUpdate(
+                        new QCallAssembly(ref rtAsm),
+                        metadataDeltaPtr,
+                        metadataDelta.Length,
+                        ilDeltaPtr,
+                        ilDelta.Length,
+                        pdbDeltaPtr,
+                        pdbDelta.Length) != 0)
+                    {
+                        throw new NotSupportedException();
```

Comment:
Is there any information we can include in the error message that would be more meaningful?

Reply:
The runtime doesn't give us much info on failures.  The pinvoke returns an HRESULT from the vm code but there really isn't any detail there. It is either going to be E_FAIL, E_INVALIDARG or E_NOINTERFACE.

Reply:
we could always change the runtime-side code to provide a more meaningful HRESULT too. 

From a priority standpoint I would guess we are expecting the Rosyln/dotnet-watch code to figure out when the edits aren't going to be successful prior to calling this method. However if there is some non-trivial slice of the scenario where dotnet-watch can't accurately predict success then the user is going to be dependent on the error information here to understand why things didn't work.

Reply:
I agree that providing detailed exception for mall-formed delta is hard and we do not need to do it.

I think it would be useful to provide more details for operator errors, e.g. throw `InvalidOperationException` with appropriate message when somebody calls this on API on assembly that is not enabled for EnC.

Note that you can throw exception in the .cpp just fine. I would change the QCall to return void and throw all exceptions from the runtime side.

Reply:
Change made the change to throw from the cpp pinvoke.